### PR TITLE
Add support for ...WithAttributes functions

### PR DESCRIPTION
**Pull request by jls5177**
Implements #41 by extending the changes from the with-attributes branch. Most of the change is just boiler plate code to the various key generation functions. The bulk of the changes is to the new helper functions and types to make working with Attributes easier. 

**Summary of new changes:**
- Converted the Attribute type into an alias for the pkcs11.Attribute type to allow for direct conversion between the 2 types. Did the same for the AttributeType type.
- Added NewAttributewith... functions to simplify the common attribute generation functions
- Created an AttributeSet type to provide some common APIs (Add, Append, Merge, Copy, AddDefaults) when working with Attributes. These APIs return a pointer to the struct to allow for chaining the methods.
- Removed the code that returned the modified Attribute slices since the new AttributeSet is mutable and will contain the added Attributes after the call.

### CODE REVIEW

File: crypto11.go [hunk 1]
```diff
@@ -310,7 +310,9 @@ func Configure(config *Config) (*Context, error) {
 		return nil, errors.WithMessagef(err, "failed to create long term session")
 	}
 	err = instance.ctx.Login(instance.persistentSession, pkcs11.CKU_USER, instance.cfg.Pin)
-	if err != nil {
+
+	// the login will fail when using a module slot so only error out when the token requires a login
+	if err != nil && (instance.token.Flags&pkcs11.CKF_LOGIN_REQUIRED) == pkcs11.CKF_LOGIN_REQUIRED {
```
Comment (dmjones) [inline crypto11.go]: This is an unrelated change, so I'd prefer we separated this out.

I'd also like to discuss the best way to approach this. I haven't tested whether CKF_LOGIN_REQUIRED is accurately populated by popular PKCS#11 implementations. An alternative might be to have a config setting controlling whether login is required or not.

File: attributes.go [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+		Attributes: map[AttributeType]*Attribute{},
+	}
+}
+
+func (a *AttributeSet) Add(attributeType AttributeType, value interface{}) *AttributeSet {
+	a.Attributes[attributeType] = NewAttribute(attributeType, value)
+	return a
+}
+
+func (a *AttributeSet) Append(attribute *Attribute) *AttributeSet {
```
Comment (dmjones) [inline attributes.go]: Given the operation of the function, would `Set` be a more appropriate name? Or something else that makes it clear an existing attribute of the same type is overridden by this.

File: attributes.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+func (a *AttributeSet) Merge(additional *AttributeSet) *AttributeSet {
+	for _, attribute := range additional.Attributes {
+		a.Attributes[attribute.Type] = attribute
+	}
+	return a
+}
+
+func (a *AttributeSet) AddDefaultAttributes(additional []*Attribute) *AttributeSet {
```
Comment (dmjones) [inline attributes.go]: The name of this function doesn't seem to match the logic. Something like `AddIfNotPresent` (or a more succinct version) might fit better.

File: attributes.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+	for _, v := range a.Attributes {
+		attributes = append(attributes, v)
+	}
+	return attributes
+}
+
+func (a *AttributeSet) Copy() *AttributeSet {
+	b := NewAttributeSet()
+	for _, v := range a.Attributes {
+		b.Attributes[v.Type] = v
```
Comment (dmjones) [inline attributes.go]: Should this be a deep copy? If not, the docs for this function will need to make it clear its shallow.

File: attributes.go [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+func NewAttributewithId(id []byte) (*AttributeSet, error) {
+	if err := notNilBytes(id, "id"); err != nil {
+		return nil, err
+	}
+	return NewAttributeSet().Add(CkaId, id), nil
+}
+
+// NewAttributewithLabel is a helper function that populates a new slice of Attributes with the provided ID and Label.
+// This function returns an error if either the ID or the Label is an empty slice.
+func NewAttributewithLabel(id, label []byte) (a *AttributeSet, err error) {
```
Comment (dmjones) [inline attributes.go]: Should this be called `NewAttributeSetWithIDAndLabel`?  (note the `Set` and `WithID`)?

File: attributes.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	pAttr := pkcs11.NewAttribute(uint(attributeType), value)
+	return &Attribute{
+		Type:  attributeType,
+		Value: pAttr.Value,
+	}
+}
+
+// NewAttributewithId is a helper function that populates a new slice of Attributes with the provided ID.
+// This function returns an error if the ID is an empty slice.
+func NewAttributewithId(id []byte) (*AttributeSet, error) {
```
Comment (dmjones) [inline attributes.go]: Should this be `NewAttributeSetWithID` (note: the `Set` as well as capitalisation)?

File: attributes.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+func NewAttributeSet() *AttributeSet {
+	return &AttributeSet{
+		Attributes: map[AttributeType]*Attribute{},
+	}
+}
+
+func (a *AttributeSet) Add(attributeType AttributeType, value interface{}) *AttributeSet {
+	a.Attributes[attributeType] = NewAttribute(attributeType, value)
```
Comment (dmjones) [inline attributes.go]: Because this function uses `NewAttribute`, we should ensure the function comments warns about the situations in which it panics.

File: attributes.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+		if err := notNilBytes(attribute.Value, fmt.Sprintf("%s id attribute", name)); err != nil {
+			return err
+		}
+		return nil
+	}
+	return fmt.Errorf("missing %s id attribute", name)
+}
+
+// keyPairAttributeHasID ensures both sets of attributes has a CkaId attribute
+func validateKeyPairAttributes(public, private *AttributeSet) error {
```
Comment (dmjones) [inline attributes.go]: This function seems to require keys to have "public" or "private" in their CKA_ID attribute. This is not something we should enforce; perhaps this is code accidentally added from your own use case?

File: dsa.go [hunk 9]
```diff
@@ -79,11 +79,11 @@ func (c *Context) GenerateDSAKeyPair(id []byte, params *dsa.Parameters) (Signer,
...
 	}
 
-	if err := notNilBytes(id, "id"); err != nil {
+	template, err := NewAttributewithId(id)
+	if err != nil {
 		return nil, err
 	}
-
-	return c.generateDSAKeyPair(id, nil, params)
+	return c.GenerateDSAKeyPairWithAttributes(template, template.Copy(), params)
```
Comment (dmjones) [inline dsa.go]: Assuming we get rid of `validateKeyPairAttributes`, there should be no need to copy the template here. (This comment applies to all similar instances in the generation functions).

File: dsa.go [hunk 10]
```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
...
 }
 
-// generateDSAKeyPair creates a DSA private key on the token.
-func (c *Context) generateDSAKeyPair(id, label []byte, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
+// GenerateDSAKeyPairWithAttributes creates a DSA key pair on the token. Additional attributes from public and
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateDSAKeyPairWithAttributes(public, private *AttributeSet, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
+	if err = validateKeyPairAttributes(public, private); err != nil {
```
Comment (dmjones) [inline dsa.go]: This function (and the other `...WithAttributes` functions) need this at the start:

```go
	if c.closed.Get() {
		return nil, errClosed
	}
```

File: dsa.go [hunk 11]
```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
...
+		return
+	}
+
 	err = c.withSession(func(session *pkcs11Session) error {
 		p := params.P.Bytes()
 		q := params.Q.Bytes()
 		g := params.G.Bytes()
 
-		publicKeyTemplate := []*pkcs11.Attribute{
+		public.AddDefaultAttributes([]*pkcs11.Attribute{
```
Comment (dmjones) [inline dsa.go]: It might be polite to take a copy of these templates before we mutate them. Or... we see this as a feature and explain in the function comments that the final applied template is "returned" in the calling arguments. What do you think?

(This comment applies to all the `...WithAttributes` functions).

File: dsa.go [hunk 12]
```diff
@@ -93,55 +93,47 @@ func (c *Context) GenerateDSAKeyPairWithLabel(id, label []byte, params *dsa.Para
...
-	return c.generateDSAKeyPair(id, label, params)
+	return c.GenerateDSAKeyPairWithAttributes(template, template.Copy(), params)
 }
 
-// generateDSAKeyPair creates a DSA private key on the token.
-func (c *Context) generateDSAKeyPair(id, label []byte, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
+// GenerateDSAKeyPairWithAttributes creates a DSA key pair on the token. Additional attributes from public and
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateDSAKeyPairWithAttributes(public, private *AttributeSet, params *dsa.Parameters) (k *pkcs11PrivateKeyDSA, err error) {
```
Comment (dmjones) [inline dsa.go]: The return values should be _unnamed_ and using the public interface type, e.g. `(Signer, error)`. This same comment applies to the other `...WithAttributes` functions.

File: symmetric.go [hunk 13]
```diff
@@ -273,26 +273,29 @@ func (c *Context) GenerateSecretKeyWithLabel(id, label []byte, bits int, cipher
...
 
 }
 
-// generateSecretKey creates an secret key of given length and type.
-func (c *Context) generateSecretKey(id, label []byte, bits int, cipher *SymmetricCipher) (k *SecretKey, err error) {
+// GenerateSecretKeyWithAttributes creates an secret key of given length and type. Additional attributes from public and
+// private will be merged into the default templates used when generating keys. Where conflicts occur, the user
+// supplied attributes will win.
+func (c *Context) GenerateSecretKeyWithAttributes(template *AttributeSet, bits int, cipher *SymmetricCipher) (k *SecretKey, err error) {
+	if err = attributeHasID(template, "secret"); err != nil {
```
Comment (dmjones) [inline symmetric.go]: Per the comments above (in `attributes.go`), I think this check ought to be removed.

File: attributes.go [hunk 14]
```diff
@@ -0,0 +1,256 @@
+package crypto11
+
+import (
+	"fmt"
+	"github.com/miekg/pkcs11"
+)
+
+type AttributeType = uint
```
Comment (dmjones) [inline attributes.go:10]: With the exception of the `CkaFoo` constants, exported types and functions must have comments. (I thought the CI checked this, but obviously not). This comment applies throughout this file (only mentioned once).

File: attributes.go [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+			a.Attributes[additionalAttr.Type] = additionalAttr
+		}
+	}
+	return a
+}
+
+func (a *AttributeSet) ToSlice() []*Attribute {
+	var attributes []*Attribute
+	for _, v := range a.Attributes {
+		attributes = append(attributes, v)
```
Comment (dmjones) [inline attributes.go]: Is a shallow copy OK here? If so, let's make that clear when adding a function comment.

**Review by dmjones — CHANGES_REQUESTED** [review]
Thank you for taking the time to finish this work. I've added quite a few review comments, but they are mostly tidy up work.

The two actual problems (which I've also added comments for) are:
- I would like to address the module-protected login issue separately
- Validation that keys have IDs with "public" and "private" is incorrect in the general use case

File: attributes.go [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+	CkaDefatCmsAttributes     = AttributeType(0x00000502)
+	CkaSupportedCmsAttributes = AttributeType(0x00000503)
+	CkaAllowedMechanisms      = AttributeType(ckfArrayAttribute | AttributeType(0x00000600))
+)
+
+// An Attribute represents a PKCS#11 CK_ATTRIBUTE type.
+type Attribute = pkcs11.Attribute
+
+// An AttributeSet groups together operations that are common for a slice of Attributes
+type AttributeSet struct {
```
Comment (dmjones) [inline attributes.go]: We could just make this a map, for simplicity.

E.g.

```go
type AttributeSet map[AttributeType]*Attribute

func NewAttributeSet() AttributeSet {
	return make(AttributeSet)
}
```

The receiver functions would need to drop from `*AttributeSet` to `AttributeSet`. What do you think?

File: dsa.go (hunk 11 above)
Comment (dmjones) [reply dsa.go]: Sorry, I've just remembered your introduction:

>Removed the code that returned the modified Attribute slices since the new AttributeSet is mutable and will contain the added Attributes after the call.

So that implies we will comment this as a feature, which I'm happy with :-)

File: crypto11.go (hunk 1 above)
Comment (jls5177) [reply crypto11.go]: I agree, I created #46 to discuss how to proceed with this issue.

File: dsa.go (hunk 12 above)
Comment (jls5177) [reply dsa.go]: I just noticed this same issue when I updated my codebase to use a version of this PR. Yes, I will change these to return the `Signer` interface instead.

**Comment by jls5177** [comment]
I appreciate all of the good feedback. I will start working on these changes and hope to have a new patchset up by the end of the week.

File: attributes.go (hunk 2 above)
Comment (jls5177) [reply attributes.go]: I agree with the name `Set()`. This will be changed in the next patchset.

File: attributes.go (hunk 3 above)
Comment (jls5177) [reply attributes.go]: Done

File: attributes.go (hunk 4 above)
Comment (jls5177) [reply attributes.go]: Yeah there is no real reason to return a shallow copy. I will update the code.

File: attributes.go (hunk 5 above)
Comment (jls5177) [reply attributes.go]: Done

File: attributes.go (hunk 6 above)
Comment (jls5177) [reply attributes.go]: Yes, I agree with the naming convention.

File: attributes.go (hunk 7 above)
Comment (jls5177) [reply attributes.go]: Makes sense, I've updated the few functions that call `NewAttribute`.

File: attributes.go (hunk 8 above)
Comment (jls5177) [reply attributes.go]: Sorry, I am still learning the PKCS#11 spec. The previous code made it seem like `CkaId` and/or `CkaLabel` was required when making these API calls. I removed the additional checks from the various `Generate...WithAttributes()` functions.

File: dsa.go (hunk 9 above)
Comment (jls5177) [reply dsa.go]: The copy is required as the `Public` attribute set has different attributes from the `Private`. If we do not peform the copy then the API will fail. I will look into refactoring the code to make this clearer.

File: dsa.go (hunk 10 above)
Comment (jls5177) [reply dsa.go]: Good catch. I will add this to the various functions.

File: dsa.go (hunk 12 above)
Comment (jls5177) [reply dsa.go]: I switched the return type to `Signer` but left it named as the bulk of the logic is within the anonymous function passed to `withSession()` and the named argument allows the value to be easily returned from the outer function.

File: symmetric.go (hunk 13 above)
Comment (jls5177) [reply symmetric.go]: Done

File: dsa.go (hunk 11 above)
Comment (jls5177) [reply dsa.go]: The original commit was trying to return the modified AttributeSets. I felt this was very hard to maintain so I decided it would be better to just update the received AttributeSet so the caller can manipulate for subsequent calls if they needed to. 

I can make a clone instead and just not return the modified `AttributeSet`. What method would you prefer?

File: attributes.go (hunk 14 above)
Comment (jls5177) [reply attributes.go:10]: I've added comments to the various types and functions.

File: attributes.go (hunk 15 above)
Comment (jls5177) [reply attributes.go]: I converted this to a deep copy to match the `Copy()` function.

File: attributes.go (hunk 16 above)
Comment (jls5177) [reply attributes.go]: Yeah this makes sense and actually makes the garbage collection faster since it does not have to track all of the pointers being passed around.

**Comment by dmjones** [comment]
Thank you for the updates. I'm afraid I won't have time to review them until at least next week. Looking forward to it though.

**Comment by dmjones** [comment]
Thanks for all your efforts. I've added a few commits on top of this, in https://github.com/ThalesIgnite/crypto11/pull/52. Which will be merged shortly.

([This StackOverflow question](https://stackoverflow.com/questions/11066388/how-can-you-add-commits-to-a-github-pull-request-of-another-person-to-your-repo) suggested this approach for adding new commits to an existing PR).
